Link error when UMF is built as a dynamic library with GLIBC < 2.34

Cannot build UMF as a shared library on the system with GLIC < 2.34

## Environment Information

- UMF version (hash commit or a tag):                         main branch
- OS(es) version(s):                                          Ubuntu 20.04
- kernel version(s):                                          5.4.0-146-generic
- compiler, libraries, and other related tools version(s):   gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0, GLIBC 2.31-0ubuntu9.15

## Please provide a reproduction of the bug:

Just try to build UMF with the following options:

```
cmake .. -DUMF_BUILD_SHARED_LIBRARY=ON
make
```

## How often bug is revealed:

always

## Actual behavior:

Build fails because of linker error:
```
[ 40%] Linking CXX executable umf_test-utils_common
/usr/bin/ld: CMakeFiles/umf_test-utils_common.dir/__/src/utils/utils_load_library.c.o: in function `util_open_library':
utils_load_library.c:(.text+0x16): undefined reference to `dlopen'
/usr/bin/ld: CMakeFiles/umf_test-utils_common.dir/__/src/utils/utils_load_library.c.o: in function `util_close_library':
utils_load_library.c:(.text+0x25): undefined reference to `dlclose'
/usr/bin/ld: CMakeFiles/umf_test-utils_common.dir/__/src/utils/utils_load_library.c.o: in function `util_get_symbol_addr':
utils_load_library.c:(.text+0x35): undefined reference to `dlsym'
collect2: error: ld returned 1 exit status
make[2]: *** [test/CMakeFiles/umf_test-utils_common.dir/build.make:197: test/umf_test-utils_common] Error 1
make[1]: *** [CMakeFiles/Makefile2:1200: test/CMakeFiles/umf_test-utils_common.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

## Expected behavior:

Build successes.

## Details

The issue happens because `umf_utils` uses dlopen. Starting GLIBC 2.34 the `libpthread` and the `libdl` are parts of glibc library (no explicit linking with libpthread and libdl is needed). But on the systems with GLIBC < 2.34 binary that uses `dlopen/dlsym/dlclose` should explicitly link with `libdl`

## Additional information about Priority and Help Requested:

Requested priority: Medium

--- a/.github/workflows/pr_push.yml
+++ b/.github/workflows/pr_push.yml
@@ -21,16 +21,19 @@ jobs:
           - os: windows-latest
             disjoint: 'OFF'
             build_tests: 'ON'
+            simple_cmake: 'OFF'
           # pure C build (Windows)
           - os: windows-latest
             disjoint: 'OFF'
             # Tests' building is off for a pure C build
             build_tests: 'OFF'
+            simple_cmake: 'OFF'
           - os: ubuntu-latest
             disjoint: 'ON'
             build_tests: 'ON'
             # Windows doesn't recognize 'CMAKE_BUILD_TYPE', it uses '--config' param in build command
             extra_build_options: '-DCMAKE_BUILD_TYPE=Release -DUMF_BUILD_BENCHMARKS=ON -DUMF_BUILD_BENCHMARKS_MT=ON'
+            simple_cmake: 'OFF'
           # pure C build (Linux)
           - os: ubuntu-latest
             disjoint: 'OFF'
@@ -38,6 +41,13 @@ jobs:
             # Tests' building is off for a pure C build
             build_tests: 'OFF'
             extra_build_options: '-DCMAKE_BUILD_TYPE=Release -DUMF_BUILD_BENCHMARKS=ON'
+            simple_cmake: 'OFF'
+          # simplest CMake
+          - os: ubuntu-latest
+            disjoint: 'OFF'
+            build_tests: 'ON'
+            extra_build_options: '-DCMAKE_BUILD_TYPE=Release'
+            simple_cmake: 'ON'
     runs-on: ${{matrix.os}}
 
     steps:
@@ -68,6 +78,7 @@ jobs:
       run: sudo bash -c "echo 0 > /proc/sys/kernel/yama/ptrace_scope"
 
     - name: Configure CMake
+      if: matrix.simple_cmake == 'OFF'
       run: >
         cmake
         -B ${{github.workspace}}/build
@@ -80,6 +91,16 @@ jobs:
         -DUMF_BUILD_TESTS=${{matrix.build_tests}}
         -DUMF_BUILD_EXAMPLES=ON
         -DUMF_BUILD_LEVEL_ZERO_PROVIDER=ON
+        -DUMF_TESTS_FAIL_ON_SKIP=ON
+        ${{matrix.extra_build_options}}
+
+    - name: Configure CMake (simple)
+      if: matrix.simple_cmake == 'ON'
+      run: >
+        cmake
+        -B ${{github.workspace}}/build
+        -DUMF_BUILD_LIBUMF_POOL_SCALABLE=ON
+        -DUMF_BUILD_SHARED_LIBRARY=ON
         -DUMF_TESTS_FAIL_ON_SKIP=ON
         ${{matrix.extra_build_options}}
 

--- a/include/umf/proxy_lib_new_delete.h
+++ b/include/umf/proxy_lib_new_delete.h
@@ -68,8 +68,8 @@ static inline void *internal_aligned_alloc(size_t alignment, size_t size) {
 #define decl_new_nothrow(n) [[nodiscard]]
 #endif // defined(_MSC_VER) && defined(_Ret_notnull_) && defined(_Post_writable_byte_size_)
 
-void operator delete(void *p) noexcept { free(p); };
-void operator delete[](void *p) noexcept { free(p); };
+void operator delete(void *p) noexcept { free(p); }
+void operator delete[](void *p) noexcept { free(p); }
 
 void operator delete(void *p, const std::nothrow_t &) noexcept { free(p); }
 void operator delete[](void *p, const std::nothrow_t &) noexcept { free(p); }
@@ -96,11 +96,11 @@ decl_new_nothrow(n) void *operator new[](std::size_t n,
 void operator delete(void *p, std::size_t n) noexcept {
     (void)(n);
     free(p);
-};
+}
 void operator delete[](void *p, std::size_t n) noexcept {
     (void)(n);
     free(p);
-};
+}
 #endif // (__cplusplus >= 201402L || _MSC_VER >= 1916)
 
 #if (__cplusplus > 201402L || defined(__cpp_aligned_new))
@@ -116,12 +116,12 @@ void operator delete(void *p, std::size_t n, std::align_val_t al) noexcept {
     (void)(n);
     (void)(al);
     free(p);
-};
+}
 void operator delete[](void *p, std::size_t n, std::align_val_t al) noexcept {
     (void)(n);
     (void)(al);
     free(p);
-};
+}
 void operator delete(void *p, std::align_val_t al,
                      const std::nothrow_t &) noexcept {
     (void)(al);

--- a/src/utils/CMakeLists.txt
+++ b/src/utils/CMakeLists.txt
@@ -32,6 +32,7 @@ endif()
 if(LINUX OR MACOSX)
     set(UMF_UTILS_SOURCES ${UMF_UTILS_SOURCES_COMMON}
                           ${UMF_UTILS_SOURCES_POSIX})
+    set(UMF_UTILS_LIBS dl)
 elseif(WINDOWS)
     set(UMF_UTILS_SOURCES ${UMF_UTILS_SOURCES_COMMON}
                           ${UMF_UTILS_SOURCES_WINDOWS})
@@ -41,7 +42,8 @@ add_library(umf_utils INTERFACE)
 add_library(${PROJECT_NAME}::utils ALIAS umf_utils)
 
 target_sources(umf_utils INTERFACE ${UMF_UTILS_SOURCES})
-target_link_libraries(umf_utils INTERFACE ${CMAKE_THREAD_LIBS_INIT})
+target_link_libraries(umf_utils INTERFACE ${CMAKE_THREAD_LIBS_INIT}
+                                          ${UMF_UTILS_LIBS})
 
 target_include_directories(
     umf_utils

